[Cheng Zhi Sheng] iP

zsh-eng opened the pull request:
Pogo
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Pogo frees your mind from handling your tasks.
It's
- text-based
- easy to learn 
- ~~FAST~~ *BLAZING* FAST

1. Download it from the [releases page](https://github.com/zsh-eng/ip/releases).
2. Double-click it
3. Add your tasks.
4. Let it manage your tasks for you 😉
And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

zsh-eng committed:
Rename chatbot

zsh-eng committed:
Add echo feature to chatbot

zsh-eng committed:
Add Task feature

zsh-eng committed:
Add the ability to check fof tasks

zsh-eng committed:
Add new task types: deadline, event, todo

zsh-eng committed:
Update UI tests

zsh-eng committed:
Add exception handling for invalid and blank tasks

zsh-eng committed:
Refactor handling input into separate method

zsh-eng committed:
Update error messages to be clearer

zsh-eng committed:
Improve error messages for invalid input

zsh-eng committed:
Add delete task function

zsh-eng committed:
Refactor actions and task types to use enums

zsh-eng committed:
Fix TextUI tests

zsh-eng committed:
Implement save tasks to disk feature

zsh-eng committed:
Implement load tasks from disk

zsh-eng committed:
Merge branch 'branch-Level-7'

zsh-eng committed:
Refactor storage logic out of Pogo class

zsh-eng committed:
Refactor code into separate packages

zsh-eng committed:
Add basic commands

zsh-eng committed:
Implement add task command

zsh-eng committed:
Add task parser for different tasks

zsh-eng committed:
Add mark and unmark tasks command

zsh-eng committed:
Add delete task command

zsh-eng committed:
Fix minor bugs in commands

zsh-eng committed:
Refactor command logic to command package

zsh-eng committed:
Update feedback to user for add task commands

zsh-eng committed:
Refactor storage logic to use TextEncoder

zsh-eng committed:
Refactor ui into TextUi class

zsh-eng committed:
Merge branch 'branch-OOP'

zsh-eng committed:
Add datetime parsing

zsh-eng committed:
Add ability to list tasks within date range

zsh-eng committed:
Merge branch 'branch-Level-8'

zsh-eng committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zsh-eng committed:
Add checkstyle linter

zsh-eng committed:
Add tests for DateTimeParser

zsh-eng committed:
Fix formatting and javadoc comments for TaskParser

zsh-eng committed:
Add test cases for TaskParser

zsh-eng committed:
Fix formatting issues for TaskTextEncoder

zsh-eng committed:
Add tests for TaskTextEncoder

zsh-eng committed:
Update gradle build for jar

zsh-eng committed:
Add JavaDocs

zsh-eng committed:
Add find command

zsh-eng committed:
Merge branch 'branch-A-JavaDoc'

zsh-eng committed:
Merge branch-Level-9 into master

bwangpj commented on src/main/java/pogo/Pogo.java:
```diff
@@ -0,0 +1,62 @@
+    private static final Storage storage = TextStorage.of();
+    private static final TextUi ui = new TextUi();
+
+    private static boolean handleInput(String input) {
```
Perhaps a javadoc comment for this method?

jonasongg replied:
Maybe it would be better to name this method something clearer to indicate it returns a boolean?

bwangpj commented on src/main/java/pogo/storage/TextStorage.java:
```diff
@@ -0,0 +1,88 @@
+        return new TextStorage();
+    }
+
+    private void createTaskFileIfNotExist() throws IOException {
```
Perhaps a javadoc comment for this method?

bwangpj commented on src/main/java/pogo/storage/TextStorage.java:
```diff
@@ -0,0 +1,88 @@
+    public List<Task> load() throws IOException {
+        createTaskFileIfNotExist();
+
+        File f = new File(TASKS_FILE);
```
Perhaps a more descriptive variable name?

bwangpj commented on src/main/java/pogo/tasks/Task.java:
```diff
@@ -0,0 +1,108 @@
+        return (isDone ? "X" : " ");
+    }
+
+    public String getStatusMessage() {
```
Perhaps a javadoc comment for this method?

bwangpj reviewed (commented):
Second PR review as per Week 4 requirements.

jonasongg commented on src/main/java/pogo/parsers/TaskParser.java:
```diff
@@ -0,0 +1,167 @@
+        final Matcher matcher = DEADLINE_PATTERN.matcher(input);
+
+        InvalidCommand ic =
+            new InvalidCommand(Messages.INVALID_TASK
```
Should these wrapped lines be indented with 8 spaces instead?

jonasongg commented on src/main/java/pogo/parsers/TaskParser.java:
```diff
@@ -0,0 +1,167 @@
+     * @param input  The input string.
+     * @param isMark Whether the command is a mark command or an unmark command.
+     */
+    public static Command parseMarkCommand(String input, boolean isMark) {
```
I like that this method name and its arguments, along with those of other methods in this class, are very clear in what they do and mean!

jonasongg commented on src/main/java/pogo/tasks/TaskTextEncoder.java:
```diff
@@ -0,0 +1,146 @@
+    private static final String SEPARATOR_ENCODING = " | ";
+    private static final String SEPARATOR_DECODING = " \\| ";
+    private static final String ID_DEADLINE = "D";
+    private static final String ID_EVENT = "E";
```
I like that you group associated constants with similar prefixes!

jonasongg commented on src/main/java/pogo/tasks/TaskTextEncoder.java:
```diff
@@ -0,0 +1,146 @@
+    /**
+     * Encodes an event task to a string.
+     * The task is encoded in the following format:
+     * E | isDone | description | from | to
```
I like the way you did the javadocs for this file; it makes it very clear what the methods do!

jonasongg commented on src/main/java/pogo/tasks/TaskTextEncoder.java:
```diff
@@ -0,0 +1,146 @@
+    @Override
+    public void visit(Event event) {
+        String eventString =
+            String.join(SEPARATOR_ENCODING, ID_EVENT, event.isDone() ? "1" : "0", event.getDescription(),
```
Might be better to indent this wrapped line with 8 spaces?

jonasongg reviewed (commented):
Overall, I think you followed the coding standard quite well!

zsh-eng commented:
Thank you @jonasongg and @bwangpj for the helpful comments! I've resolved the issues

zsh-eng committed:
Fix minor coding standard issues

zsh-eng committed:
Merge branch 'branch-A-Checkstyle'

zsh-eng committed:
Add GUI

Currently, the Pogo chatbot application uses a command-line interface
to display tasks to the user.

A graphical user interface will provide greater usability, without
compromising on the functionality of keyboard-only input.

Let's create a GUI for Pogo using JavaFX.

zsh-eng committed:
Merge branch 'branch-Level-10'

zsh-eng committed:
Add basic styling to GUI

zsh-eng committed:
Fix styling issues with DialogBox

zsh-eng committed:
Improve UI of main window

zsh-eng committed:
Add assertions

The classes in the application lack any assertions.

Adding assertions improves code quality by defining assumptions about the
program state that the runtime can verify.

Let's add assertions to all classes (where applicable).

Refer to to Oracle's guide on assertions to understand their benefits.
https://docs.oracle.com/javase/8/docs/technotes/guides/language/assert.html

zsh-eng committed:
Extract list tasks logic to TaskParser

The logic for parsing a "list" tasks command is placed in the Parser
class, which violates SLAP (Single Level of Abstraction).

Extracting the logic to the TaskParser class improves code quality.

Let's create a new parseListCommand method in TaskParser to handle the
parsing logic for "list".

zsh-eng committed:
Refactor TaskTextEncoder

There are numerous coding quality violations in the TaskTextEncoder class,
including magic constants, long methods, and violating SLAP.

Refactoring the existing code to fix these violations will improve quality
and readability.

Let's
* remove magic constants and replace them with constants with meaningful names
* extract decoding logic in switch statement to separate methods to improve
readability

zsh-eng committed:
Add comments to improve readability

zsh-eng committed:
Merge pull request #1 from zsh-eng/branch-A-CodeQuality

Improve code quality

zsh-eng committed:
Merge branch 'master'

zsh-eng committed:
Merge pull request #2 from zsh-eng/branch-A-Assertions

Add assertions

zsh-eng committed:
Add CI with GitHub Actions

zsh-eng committed:
Fix style issues

zsh-eng committed:
Merge branch 'branch-A-CI'

zsh-eng committed:
Improve search function for find task command

Find command matches based on substring.
A find command that matches partially, e.g. "Hor" matches "Homework" would
be more user-friendly as the user does not have to type out the entire
word of the task.

Let's
* update the search algorithm to allow for partial matching
* add a test suite for the next filtering algorithm

zsh-eng committed:
Merge branch 'master' into branch-Level-9

zsh-eng committed:
Fix gradle build file

zsh-eng committed:
Add user guide

zsh-eng committed:
Merge branch 'branch-A-UserGuide'

zsh-eng committed:
Add product screenshot

zsh-eng committed:
Fix bug related to double saving

A bug in TaskTextEncoder caused tasks to be double saved.

The reason for the bug was that the tasksString array was not cleared
before each call of `encode`, causing the tasksString to keep growing
longer and longer.

This commit fixes that bug.

zsh-eng committed:
Fix failing tests

An error in the way the tests are set up caused the tests to fail.

zsh-eng committed:
Add additional parsing methods for datetime

zsh-eng committed:
Fix minor UI bugs
